Modules/_sqlite/row.c

```diff
@@ -31,14 +31,27 @@ class _sqlite3.Row "pysqlite_Row *" "pysqlite_RowType"
 [clinic start generated code]*/
 /*[clinic end generated code: output=da39a3ee5e6b4b0d input=384227da65f250fd]*/
 
-static void
-pysqlite_row_dealloc(pysqlite_Row *self)
+static int
+row_clear(pysqlite_Row *self)
 {
-    PyTypeObject *tp = Py_TYPE(self);
+    Py_CLEAR(self->data);
+    Py_CLEAR(self->description);
+    return 0;
+}
 
-    Py_XDECREF(self->data);
-    Py_XDECREF(self->description);
+static int
+row_traverse(pysqlite_Row *self, visitproc visit, void *arg)
```

Comment:
Same as https://github.com/python/cpython/pull/26104/files#r638328191

Reply:
Otherwise is probably missing a visit for `self->data` and `self->description`

Reply:
Yes, I left those out because it's not possible to create a ref. cycle, but OTOH it should not hurt to visit them :)

Reply:
Also, we must visit the type.

Reply:
In general we should default to traverse everything, because is not only about forming cycles through those by setting something manually but possible links to and from other stuff by following existing links. A typical one is  an object to its class, and to its module (even built-ins). The place where it makes more sense to skip some of these is tp_clear, but is also not a lot of benefits unless the type is very performance sensitive and there is still the possibility to get it wrong.

So my advice for this PR is making sure wer are traversing and clearing everything that is a python object :)

Reply:
Got it! Maybe the dev guide should be updated to mention this. I can open a PR for it, if you agree.

Reply:
In theory we cover this already:

https://docs.python.org/3/c-api/gcsupport.html#c.traverseproc

In particular:

> Traversal function for a container object. Implementations must call the visit function for each object directly contained by self,

But I don't think it hurts to clarify it in the devguide

Reply:
Additionally we have this:

https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_traverse

> Note that Py_VISIT() is called only on those members that can participate in reference cycles. Although there is also a self->key member, it can only be NULL or a Python string and therefore cannot be part of a reference cycle.
> On the other hand, even if you know a member can never be part of a cycle, as a debugging aid you may want to visit it anyway just so the gc module’s get_referents() function will include I

In general, I still prefer to visit them, specially since if we ever switch strings to heap types fixing this is going to be a nightmare

Reply:
Good points. Thanks for your insight!